## Gather

In [466]:
import pandas as pd
import numpy as np
pd.options.display.max_columns = 50
import json
import requests
from io import StringIO
from urllib.request import Request, urlopen
import re

In [467]:
#gather data from twitter-archive-enhanced.csv
df_twitter_archive_enhanced = pd.read_csv('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv')

In [468]:
#gather data from image-predictions.tsv
url_image_predictions = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response_image_predictions = requests.get(url_image_predictions).text
df_image_predictions = pd.read_csv(StringIO(response_image_predictions), sep="\t")

In [469]:
#gather data from tweet_json.txt. I got this file using Twitter API 
with open ("tweet_json.txt") as file:
    data = [json.loads(line) for line in file]
df_tweet_json = pd.DataFrame(data)

## Assess

In [470]:
df_twitter_archive_enhanced

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [471]:
df_image_predictions

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [472]:
df_tweet_json

created_at                  id              id_str  \
0     Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1     Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2     Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3     Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4     Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   
...                              ...                 ...                 ...   
2325  Mon Nov 16 00:24:50 +0000 2015  666049248165822465  666049248165822465   
2326  Mon Nov 16 00:04:52 +0000 2015  666044226329800704  666044226329800704   
2327  Sun Nov 15 23:21:54 +0000 2015  666033412701032449  666033412701032449   
2328  Sun Nov 15 23:05:30 +0000 2015  666029285002620928  666029285002620928   
2329  Sun Nov 15 22:32:08 +0000 2015  666020888022790149  666020888022790149   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2325  Here we have a 1949 1st generation vulpix. Enj...      False   
2326  This is a purebred Piers Morgan. Loves to Netf...      False   
2327  Here is a very happy pup. Big fan of well-main...      False   
2328  This is a western brown Mitsubishi terrier. Up...      False   
2329  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2325           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2326           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2327           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2328           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2329           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2325  {'media': [{'id': 666049244999131136, 'id_str'...   
2326  {'media': [{'id': 666044217047650304, 'id_str'...   
2327  {'media': [{'id': 666033409081393153, 'id_str'...   
2328  {'media': [{'id': 666029276303482880, 'id_str'...   
2329  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2325  <a href="http://twitter.com/download/iphone" r...   
2326  <a href="http://twitter.com/download/iphone" r...   
2327  <a href="http://twitter.com/download/ipho

In [473]:
df_twitter_archive_enhanced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [474]:
df_image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [475]:
df_tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2330 entries, 0 to 2329
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   created_at                     2330 non-null   object 
 1   id                             2330 non-null   int64  
 2   id_str                         2330 non-null   object 
 3   full_text                      2330 non-null   object 
 4   truncated                      2330 non-null   bool   
 5   display_text_range             2330 non-null   object 
 6   entities                       2330 non-null   object 
 7   extended_entities              2058 non-null   object 
 8   source                         2330 non-null   object 
 9   in_reply_to_status_id          77 non-null     float64
 10  in_reply_to_status_id_str      77 non-null     object 
 11  in_reply_to_user_id            77 non-null     float64
 12  in_reply_to_user_id_str        77 non-null     o

In [476]:
all_columns = pd.Series(list(df_twitter_archive_enhanced) + list(df_image_predictions))
all_columns[all_columns.duplicated()]

17    tweet_id
dtype: object

In [477]:
list(df_twitter_archive_enhanced)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [478]:
df_twitter_archive_enhanced[df_twitter_archive_enhanced['expanded_urls'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [479]:
df_twitter_archive_enhanced.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [480]:
df_image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [481]:
df_twitter_archive_enhanced.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [482]:
df_tweet_json.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2197
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       32
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      10
Name: source, dtype: int64

In [483]:
df_twitter_archive_enhanced.name.value_counts()

None       745
a           55
Charlie     12
Cooper      11
Lucy        11
          ... 
Tuco         1
Bayley       1
Barry        1
Oddie        1
Eevee        1
Name: name, Length: 957, dtype: int64

In [484]:
df_twitter_archive_enhanced[df_twitter_archive_enhanced.name == 'a']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1193  717537687239008257                    NaN                  NaN   
1207  715733265223708672                    NaN                  NaN   
1340  704859558691414016                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1361  703079050210877440                    NaN                  NaN   
1368  702539513671897089                    NaN                  NaN   
1382  700864154249383937                    NaN                  NaN   
1499  692187005137076224                    NaN                  NaN   
1737  679530280114372609                    NaN                  NaN   
1785  677644091929329666                    NaN                  NaN   
1853  675706639471788032                    NaN                  NaN   
1854  675534494439489536                    NaN                  NaN   
1877  675109292475830276                    NaN                  NaN   
1878  675047298674663426                    NaN                  NaN   
1923  674082852460433408                    NaN                  NaN   
1941  673715861853720576                    NaN                  NaN   
1955  673636718965334016                    NaN                  NaN   
1994  672604026190569472                    NaN                  NaN   
2034  671743150407421952                    NaN                  NaN   
2066  671147085991960577                    NaN                  NaN   
2116  670427002554466305                    NaN                  NaN   
2125  670361874861563904                    NaN                  NaN   
2128  670303360680108032                    NaN                  NaN   
2146  669923323644657664                    NaN                  NaN   
2153  669661792646373376                    NaN                  NaN   
2161  669564461267722241                    NaN                  NaN   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN         

In [485]:
df_twitter_archive_enhanced[df_twitter_archive_enhanced.name.duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
23    2017-07-19 00:47:34 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
23    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
23    This is Canela. She attempted some fancy porch...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
7                          NaN                        NaN   
12                         NaN                        NaN   
23                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
23    https://twitter.com/dog_rates/status/887473957...                13   
24    https://twitt

In [486]:
#Most of the time denominator is 10 but sometimes has another number -- Score not valid
df_twitter_archive_enhanced.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [487]:
#only three denominator value are less than 10
df_twitter_archive_enhanced[df_twitter_archive_enhanced.rating_denominator < 10].rating_denominator

313     0
516     7
2335    2
Name: rating_denominator, dtype: int64

In [488]:
#numerator value are less than 10
df_twitter_archive_enhanced[df_twitter_archive_enhanced.rating_numerator < 10].rating_numerator

45      5
229     6
315     0
387     7
462     7
       ..
2351    5
2352    6
2353    9
2354    7
2355    8
Name: rating_numerator, Length: 440, dtype: int64

In [582]:
#check number of nulls of expanded_urls
df_twitter_archive_enhanced.expanded_urls.isnull().value_counts()

False    2297
True       59
Name: expanded_urls, dtype: int64

In [583]:
df_twitter_archive_enhanced[df_twitter_archive_enhanced.expanded_urls.isnull() == True]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [584]:
#verify rows where rating_numerator is less than 10
df_twitter_archive_enhanced[df_twitter_archive_enhanced.rating_numerator < 10].head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45   883482846933004288                    NaN                  NaN   
229  848212111729840128                    NaN                  NaN   
315  835152434251116546                    NaN                  NaN   
387  826598799820865537           8.265984e+17         4.196984e+09   
462  817502432452313088                    NaN                  NaN   

                     timestamp  \
45   2017-07-08 00:28:19 +0000   
229  2017-04-01 16:35:01 +0000   
315  2017-02-24 15:40:31 +0000   
387  2017-02-01 01:11:25 +0000   
462  2017-01-06 22:45:43 +0000   

                                                source  \
45   <a href="http://twitter.com/download/iphone" r...   
229  <a href="http://twitter.com/download/iphone" r...   
315  <a href="http://twitter.com/download/iphone" r...   
387  <a href="http://twitter.com/download/iphone" r...   
462  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
45   This is Bella. She hopes her smile made you sm...                  NaN   
229  This is Jerry. He's doing a distinguished tong...                  NaN   
315  When you're so blinded by your systematic plag...                  NaN   
387  I was going to do 007/10, but the joke wasn't ...                  NaN   
462  RT @dog_rates: Meet Herschel. He's slightly bi...         6.924173e+17   

     retweeted_status_user_id retweeted_status_timestamp  \
45                        NaN                        NaN   
229                       NaN                        NaN   
315                       NaN                        NaN   
387                       NaN                        NaN   
462              4.196984e+09  2016-01-27 18:42:06 +0000   

                                         expanded_urls  rating_numerator  \
45   https://twitter.com/dog_rates/status/883482846...                 5   
229  https://twitter.com/dog_rates/status/848212111...                 6   
315  https://twitter.com/dog_rates/status/835152434...                 0   
387                                                NaN                 7   
462  https://twitter.com/dog_rates/status/692417313...                 7   

     rating_denominator      name doggo floofer  pupper puppo  
45                   10     Bella  None    None    None  None  
229                  10     Jerry  None    None    None  None  
315                  10      None  None    None    None  None  
387                  10      None  None    None    None  None  
462                  10  Herschel  None    None  pupper  None

In [585]:
#it can be seen that some text range start with 12 but if I count the characters the value value is the second
df_tweet_json.display_text_range.value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[0, 140]    251
[0, 139]    145
[0, 138]     79
[0, 137]     77
[0, 116]     59
           ... 
[12, 56]      1
[10, 28]      1
[0, 21]       1
[0, 145]      1
[0, 148]      1
Name: display_text_range, Length: 143, dtype: int64

In [587]:
#making sure there are no values after 1 August 2017
df_twitter_archive_enhanced[df_twitter_archive_enhanced.timestamp > '2017-08-01 00:28:19 +0000'].head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [588]:
#Check that there are tweet text that start by 13 display_text_range 
df_tweet_json[df_tweet_json.display_text_range.str[0] == 13]

created_at                  id              id_str  \
54    Sun Jul 02 21:58:53 +0000 2017  881633300179243008  881633300179243008   
328   Thu Feb 16 04:45:50 +0000 2017  832088576586297345  832088576586297345   
558   Tue Nov 22 00:32:18 +0000 2016  800859414831898624  800859414831898624   
1056  Sat Jun 04 00:32:32 +0000 2016  738891149612572673  738891149612572673   

                                              full_text  truncated  \
54    @roushfenway These are good dogs but 17/10 is ...      False   
328            @docmisterio account started on 11/15/15      False   
558   @SkyWilliams doggo simply protecting you from ...      False   
1056                                 @mount_alex3 13/10      False   

     display_text_range                                           entities  \
54             [13, 91]  {'hashtags': [], 'symbols': [], 'user_mentions...   
328            [13, 40]  {'hashtags': [], 'symbols': [], 'user_mentions...   
558           [13, 105]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1056           [13, 18]  {'hashtags': [], 'symbols': [], 'user_mentions...   

     extended_entities                                             source  \
54                 NaN  <a href="http://twitter.com/download/iphone" r...   
328                NaN  <a href="http://twitter.com/download/iphone" r...   
558                NaN  <a href="http://twitter.com/download/iphone" r...   
1056               NaN  <a href="http://twitter.com/download/iphone" r...   

      in_reply_to_status_id in_reply_to_status_id_str  in_reply_to_user_id  \
54             8.816070e+17        881607037314052101           47384430.0   
328            8.320875e+17        832087547559997440           30582082.0   
558            8.008580e+17        800857954417262593          291859009.0   
1056           7.384119e+17        738411919896285185          358972768.0   

     in_reply_to_user_id_str in_reply_to_screen_name  \
54                  47384430             roushfenway   
328                 30582082             docmisterio   
558                291859009             SkyWilliams   
1056               358972768             mount_alex3   

                                                   user   geo coordinates  \
54    {'id': 4196983835, 'id_str': '4196983835', 'na...  None        None   
328   {'id': 4196983835, 'id_str': '4196983835', 'na...  None        None   
558   {'id': 4196983835, 'id_str': '4196983835', 'na...  None        None   
1056  {'id': 4196983835, 'id_str': '4196983835', 'na...  None        None   

     place contributors  is_quote_status  retweet_count  favorite_count  \
54    None         None            False              7             115   
328   None         None            False              2              61   
558   None         None            False             96             677   
1056  None         None            False              6             102   

      favorited  retweeted possibly_sensitive possibly_sensitive_appealable  \
54        False      False                NaN                           NaN   
328       False      False                NaN                           NaN   
558       False      False                NaN                           NaN   
1056      False      False                NaN                           NaN   

     lang retweeted_status  quoted_status_id quoted_status_id_str  \
54     en              NaN               NaN                  NaN   
328    en              NaN               NaN                  NaN   
558    en              NaN               NaN                  NaN   
1056  und              NaN               NaN                  NaN   

     quoted_status_permalink quoted_status  
54                       NaN           NaN  
328                      NaN           NaN  
558                      NaN           NaN  
1056                     NaN           NaN

In [589]:
#it can gather the first part of the range of display_text_range
df_tweet_json.display_text_range.str[0].value_counts()

0     2307
13       4
12       3
5        2
16       2
14       2
27       1
23       1
17       1
15       1
7        1
30       1
28       1
22       1
18       1
10       1
Name: display_text_range, dtype: int64

In [590]:
#retweeted only shows False values
df_tweet_json.retweeted.value_counts()

False    2330
Name: retweeted, dtype: int64

In [591]:
#checking the field user which is a json
df_tweet_json.user[:][0]

{'id': 4196983835,
 'id_str': '4196983835',
 'name': 'WeRateDogs®',
 'screen_name': 'dog_rates',
 'location': 'links and things ➜',
 'description': 'Your Only Source For Professional Dog Ratings Instagram and Facebook ➜ WeRateDogs partnerships@weratedogs.com ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀',
 'url': 'https://t.co/6ytGi24QCk',
 'entities': {'url': {'urls': [{'url': 'https://t.co/6ytGi24QCk',
     'expanded_url': 'http://campsite.bio/weratedogs',
     'display_url': 'campsite.bio/weratedogs',
     'indices': [0, 23]}]},
  'description': {'urls': []}},
 'protected': False,
 'followers_count': 8964295,
 'friends_count': 17,
 'listed_count': 6371,
 'created_at': 'Sun Nov 15 21:41:29 +0000 2015',
 'favourites_count': 145799,
 'utc_offset': None,
 'time_zone': None,
 'geo_enabled': True,
 'verified': True,
 'statuses_count': 13464,
 'lang': None,
 'contributors_enabled': False,
 'is_translator': False,
 'is_translation_enabled': False,
 'profile_background_color': '000000',
 'profile_background_image_url': 'htt

In [498]:
#checking that all user id are the same
new_array = []
for i in df_tweet_json['user']:
    print(i['id'])
    new_array.append(i['id'])

4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835

4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835
4196983835

In [499]:
result = new_array.count(new_array[0]) == len(new_array)
if result:
    print("all elements are equal")

all elements are equal


#### Quality
##### `df_twitter_archive_enhanced` table
- Delete columns with mostly null values: in_reply_to_status_id, in _reply_to_userr_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp
- expanded_urls only 2297 not nulls
- timestamp is an object string and should be a date
- source is dirty it has a tag and href and all seem to be the same
- Create dog_stage a categorical variable for doggo, floofer, pupper, puppo and None it does not have a value. 
- There are  745 dogs named a
- Canela name is repeated
- May be change rating denominator and numerator to only one column
- Numerator and denominator ratings should not be less than 10
- Clean source and leave only the name of the source


##### `df_image_predictions` table
- Some capitalized and other lower case in columns p1, p2 and p3
- Missing record 2075 instead of 2356
- I don't understand img_num sometimes 1 and 2 

##### `df_tweet_json` table
- Change the column name id to tweet_id as the other tables
- id_str seem to be redundant
- source check if it is the same for all then delete as it doesn't add much information
- only 2330 rows
- delete columns: in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, in_reply_to_screen_name, geo, coordinates, place, contributors, retweeted_status, quoted_status_id, quoted_status_permalink, quoted_status  
- delete source duplicated column

#### Tidiness
- create only one column for rating
- retweet_count_and favorite_count should be in `df_tweeter_archive_enhanced` table


## Clean

In [551]:
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced.copy()
df_image_predictions_clean = df_image_predictions.copy()
df_tweet_json_clean = df_tweet_json.copy()

### Redundant Data

#### `df_tweeter_archive_enhanced`: Delete columns with mostly null values

##### Define
Delete these columns : in_reply_to_status_id, in _reply_to_userr_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp.

##### Code

In [552]:
df_twitter_archive_enhanced_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [553]:
df_twitter_archive_enhanced_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True, axis=1)


##### Test

In [554]:
df_twitter_archive_enhanced_clean.head(1)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

#### `df_twitter_archive_enhanced`: timestamp is an object string and should be a date

##### Define
change `timestamp` column: from `string` to `date` type. 

##### Code

In [555]:
df_twitter_archive_enhanced_clean['timestamp'] = pd.to_datetime(df_twitter_archive_enhanced_clean['timestamp'])

##### Test

In [556]:
df_twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2356 non-null   int64              
 1   timestamp           2356 non-null   datetime64[ns, UTC]
 2   source              2356 non-null   object             
 3   text                2356 non-null   object             
 4   expanded_urls       2297 non-null   object             
 5   rating_numerator    2356 non-null   int64              
 6   rating_denominator  2356 non-null   int64              
 7   name                2356 non-null   object             
 8   doggo               2356 non-null   object             
 9   floofer             2356 non-null   object             
 10  pupper              2356 non-null   object             
 11  puppo               2356 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(3

#### `df_twitter_archive_enhanced`: Clean source and leave only the name of the source

##### Define
in `source` remove the tags and leave only the names. 

##### Code

In [557]:
df_twitter_archive_enhanced_clean.source[0]

'<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>'

In [558]:
df_twitter_archive_enhanced_clean.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [559]:
df_twitter_archive_enhanced_clean['source'] = df_twitter_archive_enhanced_clean.source.str.extract('>([^;]*)<')

##### Test

In [560]:
df_twitter_archive_enhanced_clean.source.head()

0    Twitter for iPhone
1    Twitter for iPhone
2    Twitter for iPhone
3    Twitter for iPhone
4    Twitter for iPhone
Name: source, dtype: object

#### `df_twitter_archive_enhanced_clean`: Create dog_stage a categorical variable for doggo, floofer, pupper, puppo and None it does not have a value.

##### Define
create a new variable dog_stage and delete all other columns: doggo, floofer, pupper, puppo

##### Code

In [561]:
#create an array with all dog stages
dog_stage_array = []
for x in range(df_twitter_archive_enhanced_clean['doggo'].size):
    if df_twitter_archive_enhanced_clean['doggo'].iloc[x] != 'None':
        dog_stage_array.append('doggo')
    elif df_twitter_archive_enhanced_clean['floofer'].iloc[x] != 'None':
        dog_stage_array.append('floofer')
    elif df_twitter_archive_enhanced_clean['pupper'].iloc[x] != 'None':
        dog_stage_array.append('pupper')
    elif df_twitter_archive_enhanced_clean['puppo'].iloc[x] != 'None':
        dog_stage_array.append('puppo')
    else :
        dog_stage_array.append('None')

In [562]:
#create new column and add it to the dataframe
df_twitter_archive_enhanced_clean['dog_stage'] = pd.Series(dog_stage_array, index=df_twitter_archive_enhanced.index)

# To category
df_twitter_archive_enhanced_clean.dog_stage = df_twitter_archive_enhanced_clean.dog_stage.astype('category')

In [563]:
#delete unnecessary columns doggo, floofer, pupper, puppo
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1) 

##### Test

In [564]:
df_twitter_archive_enhanced_clean.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name dog_stage  
0                  10   Phineas      None  
1                  10     Tilly      None  
2                  10    Archie      None  
3                  10     Darla      None  
4                  10  Franklin      None

In [565]:
df_twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2356 non-null   int64              
 1   timestamp           2356 non-null   datetime64[ns, UTC]
 2   source              2356 non-null   object             
 3   text                2356 non-null   object             
 4   expanded_urls       2297 non-null   object             
 5   rating_numerator    2356 non-null   int64              
 6   rating_denominator  2356 non-null   int64              
 7   name                2356 non-null   object             
 8   dog_stage           2356 non-null   category           
dtypes: category(1), datetime64[ns, UTC](1), int64(3), object(4)
memory usage: 149.9+ KB


### Tidiness

#### In `df_twitter_archive_enhanced_clean` create only one column for ratings 

##### Define
Delete denominator column as most ratings are 10 and only three denominator values are less than 10. We will leave numerator ratings but more than 440 are less than 10 we will analyze the effect of droping them in the analysis section. As there will be only one column for rating we will change rating_numerator to rating only.

##### Code

In [566]:
# Note: axis=1 denotes that we are referring to a column, not a row
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean.drop('rating_denominator', axis=1)

#rename rating_numerator to rating
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean.rename(columns={'rating_numerator': 'rating'})

##### Test

In [567]:
# Confirm contact column is gone
df_twitter_archive_enhanced_clean.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating      name  \
0  https://twitter.com/dog_rates/status/892420643...      13   Phineas   
1  https://twitter.com/dog_rates/status/892177421...      13     Tilly   
2  https://twitter.com/dog_rates/status/891815181...      12    Archie   
3  https://twitter.com/dog_rates/status/891689557...      13     Darla   
4  https://twitter.com/dog_rates/status/891327558...      12  Franklin   

  dog_stage  
0      None  
1      None  
2      None  
3      None  
4      None

#### In columns p1, p2 and p3 in `df_image_predictions_clean` some capitalized and others lower case

##### Define
Let's lower case p1, p2 and p3 so that all are even

##### Code

In [568]:
df_image_predictions_clean[['p1', 'p2', 'p3']] = df_image_predictions_clean[['p1', 'p2', 'p3']].astype(str).apply(lambda x: x.str.lower())

##### Test

In [569]:
df_image_predictions_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         german_shepherd  0.596461    True            malinois   
3        1     rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    shetland_sheepdog  0.061428    True  
1  0.074192    True  rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             doberman  0.154629    True

#### In `df_tweet_json_clean` table change column name `id` to `tweet_id` as the other tables

##### Define
Change clumn name `id` to `tweet_id`.

##### Code

In [570]:
df_tweet_json_clean = df_tweet_json_clean.rename(columns={'id': 'tweet_id'})

##### Test

In [571]:
df_tweet_json_clean.head(1)

created_at            tweet_id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   

  in_reply_to_status_id_str  in_reply_to_user_id in_reply_to_user_id_str  \
0                      None                  NaN                    None   

  in_reply_to_screen_name                                               user  \
0                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   

    geo coordinates place contributors  is_quote_status  retweet_count  \
0  None        None  None         None            False           7373   

   favorite_count  favorited  retweeted possibly_sensitive  \
0           35072      False      False              False   

  possibly_sensitive_appealable lang retweeted_status  quoted_status_id  \
0                         False   en              NaN               NaN   

  quoted_status_id_str quoted_status_permalink quoted_status  
0                  NaN                     NaN           NaN

#### display_text_range in `df_tweet_json` table is not uselful as a list better change to numerical variable and also change the column name to something more descriptive text_length

##### Define
pickup the right part of the range of list as the text length and rename column name to text_length

##### Code

In [572]:
df_tweet_json_clean.display_text_range = df_tweet_json_clean.display_text_range.str[1]
df_tweet_json_clean = df_tweet_json_clean.rename(columns = {'display_text_range' : 'text_length'})

##### Test

In [573]:
df_tweet_json_clean.head()

created_at            tweet_id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1  Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2  Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3  Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4  Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   

                                           full_text  truncated  text_length  \
0  This is Phineas. He's a mystical boy. Only eve...      False           85   
1  This is Tilly. She's just checking pup on you....      False          138   
2  This is Archie. He is a rare Norwegian Pouncin...      False          121   
3  This is Darla. She commenced a snooze mid meal...      False           79   
4  This is Franklin. He would like you to stop ca...      False          138   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   
2  {'media': [{'id': 891815175371796480, 'id_str'...   
3  {'media': [{'id': 891689552724799489, 'id_str'...   
4  {'media': [{'id': 891327551943041024, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   
1  <a href="http://twitter.com/download/iphone" r...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...                    NaN   
3  <a href="http://twitter.com/download/iphone" r...                    NaN   
4  <a href="http://twitter.com/download/iphone" r...                    NaN   

  in_reply_to_status_id_str  in_reply_to_user_id in_reply_to_user_id_str  \
0                      None                  NaN                    None   
1                      None                  NaN                    None   
2                      None                  NaN                    None   
3                      None                  NaN                    None   
4                      None                  NaN                    None   

  in_reply_to_screen_name                                               user  \
0                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   
1                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   
2                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   
3                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   
4                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   

    geo coordinates place contributors  is_quote_status  retweet_count  \
0  None        None  None         None            False           7373   
1  None        None  None         None            False           5494   
2  None        None  None         None            False           3632   
3  None        None  None         None            False           7550   
4  None        None  None         None            False           8137   

   favorite_count  favorited  retweeted possibly_sensitive  \
0           35072      False      False              False   
1           30366      False      False              False   
2           22829      False      False              False   
3           38342      False      False              False   
4           36622      False      False              False   

  possibly_sensitive_appealable lang retweeted_status  quoted_status_id  \
0                         False   en              NaN               NaN   
1                       

### Quality

#### Delete duplicated column, full_text source from tweet_json table. Also get rid of id_str which is the same as tweet_id but in string. It is the same information in other format. Delete also created_at, truncated, entities, extended_entities they show duplicated information. Delete because their value is the same for all rows, it is none of nan: in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, in_reply_to_screen_name, geo, coordinates, place, contributors, is_quote_status, favorited, retweeted, possibly_sensitive, possibly_sensitive_appealable, lang, user,retweeted_status, quoted_status_id, quoted_status_id_str, quoted_status_permalink, quoted_status

##### Define
using drop to delete all redundant columns mentioned above

##### Code

In [574]:
df_tweet_json_clean = df_tweet_json_clean.drop(['created_at','source','id_str','full_text','truncated' ,'entities','user','extended_entities', 'in_reply_to_status_id', 'in_reply_to_status_id_str', 'in_reply_to_user_id', 'in_reply_to_user_id_str', 'in_reply_to_screen_name', 'geo', 'coordinates', 'place', 'contributors', 'is_quote_status', 'favorited', 'retweeted', 'possibly_sensitive', 'possibly_sensitive_appealable', 'lang', 'retweeted_status', 'quoted_status_id', 'quoted_status_id_str', 'quoted_status_permalink', 'quoted_status'], axis=1) 

##### Test

In [575]:
df_tweet_json_clean.head(1)

tweet_id  text_length  retweet_count  favorite_count
0  892420643555336193           85           7373           35072

### Tidiness

#### create master archive

##### Define
Merge the three tables(df_twitter_archive_enhanced_clean,df_image_predictions_clean ,df_tweet_json_clean) into one master data file using merge 'twitter_archive_master.csv'

##### Code

In [576]:
merged1 = pd.merge(left=df_twitter_archive_enhanced_clean, right=df_image_predictions_clean, on="tweet_id")

In [577]:
merged2 = pd.merge(left=merged1, right=df_tweet_json_clean, on="tweet_id")

##### Test

In [578]:
merged2.head(1)

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls  rating     name  \
0  https://twitter.com/dog_rates/status/892420643...      13  Phineas   

  dog_stage                                          jpg_url  img_num      p1  \
0      None  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1  orange   

    p1_conf  p1_dog     p2   p2_conf  p2_dog      p3  p3_conf  p3_dog  \
0  0.097049   False  bagel  0.085851   False  banana  0.07611   False   

   text_length  retweet_count  favorite_count  
0           85           7373           35072

#### Create master archive 

In [579]:
# storing the data in csv format 
merged2.to_csv(r'twitter_archive_master.csv', index = False)

In [580]:
merged2.head(1)

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls  rating     name  \
0  https://twitter.com/dog_rates/status/892420643...      13  Phineas   

  dog_stage                                          jpg_url  img_num      p1  \
0      None  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1  orange   

    p1_conf  p1_dog     p2   p2_conf  p2_dog      p3  p3_conf  p3_dog  \
0  0.097049   False  bagel  0.085851   False  banana  0.07611   False   

   text_length  retweet_count  favorite_count  
0           85           7373           35072

## Summary

Quality

1.Delete columns unnecessary columns because they do not add value to the analysis

2.df_twitter_archive_enhanced: timestamp is an object string and should be a date

3.df_twitter_archive_enhanced: Clean source and leave only the name of the source

4.create a new variable dog_stage and delete all other columns: doggo, floofer, pupper, puppo

5.display_text_range in df_tweet_json table only use the end as that counts the number of characters

6.In columns p1, p2 and p3 in df_image_predictions_clean some capitalized and others lower case

7.In df_tweet_json_clean table change column name id to tweet_id as the other tables

8.Change the column name to something more descriptive text_length

Tidiness

1.create master archive - Merge the three tables(df_twitter_archive_enhanced_clean,df_image_predictions_clean ,df_tweet_json_clean) into one master data file using merge 'twitter_archive_master.csv'

2.create only one column for rating 